### PR TITLE
Union inputs

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: src/server/pps/server/datum_set.go [hunk 1]
```diff
@@ -1,94 +1,133 @@
...
-	for i := range d.indexes {
-		d.indexes[i] = 0
+func (d *unionDatumFactory) Datum(i int) []*workerpkg.Input {
+	for _, datumFactory := range d.inputs {
+		if i < datumFactory.Len() {
+			return datumFactory.Datum(i)
+		}
+		i -= datumFactory.Len()
 	}
+	panic("index out of bounds")
```
Comment (derekchiang) [inline src/server/pps/server/datum_set.go]: Do we want to `panic`?  I think panicing is fine if this is a truly exceptional, unrecoverable case, but just wanted to double check.

File: src/server/pps/server/datum_set.go [hunk 2]
```diff
@@ -1,94 +1,133 @@
 package server
 
 import (
```
Comment (derekchiang) [inline src/server/pps/server/datum_set.go]: I like the abstractions in this file.

**Comment by derekchiang** [comment]
Also, this PR seems to lack `pachctl` updates.  Will `inspect-job` print `input` correctly?

File: src/server/pps/server/datum_set.go (hunk 1 above)
Comment (jdoliner) [inline src/server/pps/server/datum_set.go]: I think it's fairly reasonable to panic here, just because that's the behavior with slices and other datum factories if you access out of bounds. The other option would be to return `nil` or an `error` which seems more complicated.

File: src/server/pachyderm_bench_test.go [hunk 3]
```diff
@@ -312,10 +309,7 @@ func benchmarkDataShuffle(b *testing.B, numTarballs int, numFilesPerTarball int,
 				Strategy: pps.ParallelismSpec_CONSTANT,
 				Constant: 4,
 			},
-			[]*pps.PipelineInput{{
-				Repo: client.NewRepo(dataRepo),
-				Glob: "/*",
-			}},
+			client.NewAtomInput("", dataRepo, "", "/*", false, ""),
```
Comment (jdoliner) [inline src/server/pachyderm_bench_test.go]: Yeah, I agree, will change.

File: src/server/pps/pretty/pretty.go [hunk 4]
```diff
@@ -47,22 +47,14 @@ func PrintJobInfo(w io.Writer, jobInfo *ppsclient.JobInfo) {
 func PrintPipelineHeader(w io.Writer) {
 	// because STATE is a colorful field it has to be at the end of the line,
 	// otherwise the terminal escape characters will trip up the tabwriter
-	fmt.Fprint(w, "NAME\tINPUT\tOUTPUT\tSTATE\t\n")
```
Comment (derekchiang) [inline src/server/pps/pretty/pretty.go]: I feel like input is a pretty important piece of information, certainly more so than creation time.  Can we just print the set of repos that are used as inputs, regardless of atom/union/cross?  If you wanna do that in another PR that's OK too just make sure to open an issue for it.

File: src/server/pps/pretty/pretty.go [hunk 5]
```diff
@@ -164,7 +156,7 @@ Parallelism Spec: {{.ParallelismSpec}}
 	CPU: {{ .ResourceSpec.Cpu }}
 	Memory: {{ .ResourceSpec.Memory }} {{end}}
 Inputs:
-{{pipelineInputs .}}
```
Comment (derekchiang) [inline src/server/pps/pretty/pretty.go]: What about `inputs`?  If we are ok with people using pipelines that have `inputs` but not `input`, then I think we should support printing `inputs` as well.

Comment (jdoliner) [inline src/server/pps/pretty/pretty.go]: When people use `inputs` they get translated into the `input` field so no matter what `input` should reflect their pipeline and give them an accurate impression of what's going on.

Although I think I should change the word `Inputs` to `Input` here.

File: src/server/pps/pretty/pretty.go (hunk 4 above)
Comment (jdoliner) [inline src/server/pps/pretty/pretty.go]: What about something like `Repo1 X (Repo2 U Repo3)` it seems like that would be a pretty readable shorthand.
